man/tss2_gettpm2object.1.md
% tss2_gettpm2object(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_gettpm2object**(1)

# SYNOPSIS

**tss2_gettpm2object** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_gettpm2object**(1) -  With this command for FAPI objects context
files which can be used by tpm2 tool commands can be created.
For persistent object only the textual representation of the handle number as
hex number will be written and for keys a tpm2 tool context file will
be written.
If the default TCTI differs from the FAPI profile the default the tcti can
be defined with the -T (--tcti) option.
**Note** To avoid wrong nv_written state in keystore before writing data
to the NV ram with tpm2_nvwrite, at least an empty string should be
written with tss2_nvwrite.

# OPTIONS

These are the available options:

  * **-f**, **\--force**:

    Force overwriting the output file.

  * **-p**, **\--path**=_STRING_:

    Path of the object for which the application data will be loaded.

  * **-c**, **\--context**=_FILENAME_ or _-_ (for stdout):

     The returned key context or handle.

[common tss2 options](common/tss2-options.md)

# EXAMPLES
```
tss2_gettpm2object --path=/HS/SRK/myRSACrypt --key-context=mykey.ctx
tss2_gettpm2object --path=/nv/Owner/mynv -c-
```
The command can be used in options of tpm2 commands:

```
handle=$(tss2_gettpm2object --path=/nv/Owner/mynv -c-)
tpm2_nvread $handle

```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_gettpm2object.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <tss2/tss2_tctildr.h>
#include <tss2/tss2_esys.h>
#include <tss2/tss2_mu.h>

#include "tools/fapi/tss2_template.h"
#include "lib/tpm2.h"
#include "lib/files.h"

/* Context struct used to store passed commandline parameters */
static struct cxt {
    char const *data;
    char const *path;
    char const *tcti;
    bool        overwrite;
} ctx;

/* Parse commandline parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'c':
        ctx.data = value;
        break;
    case 'f':
        ctx.overwrite = true;
        break;
    case 'p':
        ctx.path = value;
        break;
    }
    return true;
}

/* Define possible commandline parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"path", required_argument, NULL, 'p'},
        {"context",required_argument, NULL, 'c'},
        {"force" , no_argument, NULL, 'f'},
        {"tcti", required_argument, NULL, 'T'},

    };
    return (*opts = tpm2_options_new ("c:fp:", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    uint8_t blob_type;
    uint8_t *esys_blob = NULL;
    size_t esys_blob_size;
    ESYS_CONTEXT *esys_ctx = NULL;
    ESYS_TR esys_handle = ESYS_TR_NONE;
    ESYS_TR esys_handle_deser = ESYS_TR_NONE;
    FILE *stream = NULL;
    TPM2_HANDLE tpm_handle;
    TSS2_RC e_rc;
    TSS2_TCTI_CONTEXT *tcti = NULL;
    tool_rc t_rc;

    /* Check availability of required parameters */
    if (!ctx.path) {
        fprintf (stderr, "path is missing, use --path\n");
        return -1;
    }

    if (!ctx.data) {
        ctx.data = "-";
    }

    /* Execute FAPI command with passed arguments */
    TSS2_RC r = Fapi_GetEsysBlob (fctx, ctx.path, &blob_type, &esys_blob, &esys_blob_size);
    if (r != TSS2_RC_SUCCESS) {
        LOG_PERR ("Fapi_GetAppData", r);
        return 1;
    }

    if (strcmp(ctx.data, "-")) {
        if (!ctx.overwrite) {
            FILE *fp = fopen(ctx.data, "rb");
            if (fp) {
                fclose(fp);
                LOG_ERR("Path: %s already exists. Please rename or delete the file!\n",
                ctx.data);
                goto error;
            }
        }
        stream = fopen(ctx.data, "w+b");
        if (!stream) {
            LOG_ERR("Could not open path \"%s\", due to error: \"%s\"", ctx.data,
                    strerror(errno));
            goto error;
        }
    } else {
        stream = stdout;
    }

    r = Fapi_GetTcti(fctx, &tcti);
    if (r != TSS2_RC_SUCCESS) {
        LOG_PERR ("Fapi_GetTcti", r);
        return 1;
    }

    e_rc = Esys_Initialize(&esys_ctx, tcti, NULL);
  
    if (blob_type == FAPI_ESYSBLOB_CONTEXTLOAD) {
        size_t offset = 0;
        TPMS_CONTEXT context;
      
        if (e_rc != TPM2_RC_SUCCESS) {
            LOG_PERR("Esys_Initialize", e_rc);
            goto error;
        }
         e_rc = Tss2_MU_TPMS_CONTEXT_Unmarshal(esys_blob, esys_blob_size, &offset, &context);
         if (e_rc != TPM2_RC_SUCCESS) {
             LOG_PERR("Tss2_MU_TPMS_CONTEXT_Unmarshal", e_rc);
             goto error;
         }
         e_rc = Esys_ContextLoad(esys_ctx, &context, &esys_handle);
         if (e_rc != TPM2_RC_SUCCESS) {
             LOG_PERR("Esys_ContextLoad", e_rc);
             goto error;
         }
         t_rc = files_save_tpm_context_to_file(esys_ctx, esys_handle, stream);
         if (t_rc != tool_rc_success) {
             goto error;
         }
         Esys_FlushContext(esys_ctx, esys_handle);
         Esys_Finalize(&esys_ctx);
    } else {
        t_rc = tpm2_tr_deserialize(esys_ctx, esys_blob, esys_blob_size, &esys_handle_deser);
        if (t_rc != tool_rc_success) {
             goto error;
        }
        e_rc = Esys_TR_GetTpmHandle(esys_ctx, esys_handle_deser, &tpm_handle);
        if (e_rc != TSS2_RC_SUCCESS) {
            LOG_PERR("Esys_TR_GetTpmHandle", e_rc);
            goto error;
        }
        int bytes_written = fprintf(stream,"0x%08x", tpm_handle);
        if (bytes_written != 10) {
            LOG_ERR("IO error for path \"%s\"", ctx.data);
            goto error;
        }
                
        Esys_TR_Close(esys_ctx, &esys_handle_deser);
        Esys_Finalize(&esys_ctx);
    }

    /* Free allocated variables */
    Fapi_Free (esys_blob);
    if (stream && stream != stdout) {
        fclose(stream);
    }

    return 0;

 error:
    if (stream && stream != stdout) {
        fclose(stream);
    }
    Fapi_Free (esys_blob);
    if (esys_handle != ESYS_TR_NONE) {
        Esys_FlushContext(esys_ctx, esys_handle);
    }
    if (esys_handle_deser != ESYS_TR_NONE) {
        Esys_TR_Close(esys_ctx, &esys_handle_deser);
    }
    if (esys_ctx) {
        Esys_Finalize(&esys_ctx);
    }
    if (tcti) {
        Tss2_TctiLdr_Finalize(&tcti);
    }
    return 1;
}

TSS2_TOOL_REGISTER("gettpm2object", tss2_tool_onstart, tss2_tool_onrun, NULL)
